button-merchant/src/main/java/com/usebutton/merchant/SSLManagerImpl.java

```diff
@@ -0,0 +1,184 @@
+        boolean hasDiffProv = ((SSLManagerImpl) instance).provider != provider;
+        if (hasDiffProv || hasDiffPass) {
+            instance = new SSLManagerImpl(provider, password);
+        }
+
+        return instance;
+    }
+
+    @VisibleForTesting
+    SSLManagerImpl(@NonNull CertificateProvider provider, @Nullable char[] password) {
+        this.password = password;
+        this.provider = provider;
+
+        init(provider);
+    }
+
+    @Override
+    public KeyStore getKeyStore()
+            throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException,
+            KeyManagementException {
+        checkExceptions();
+        return keyStore;
+    }
+
+    @Override
+    public SSLContext getSecureContext()
+            throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException,
+            KeyManagementException {
+        checkExceptions();
+        return sslContext;
+    }
+
+    @Override
+    public CertificateProvider getCertificateProvider() {
+        return provider;
+    }
+
+    List<Certificate> getCertificateChain() {
+        return certificates;
+    }
+
+    char[] getPassword() {
+        return password;
+    }
+
+    KeyStore getKeyStore(CertificateProvider provider)
+            throws CertificateException, KeyStoreException, IOException, NoSuchAlgorithmException {
+        KeyStore keyStore = KeyStore.getInstance(KeyStore.getDefaultType());
+        keyStore.load(null, password);
+
+        certificates = provider.getChain();
+        if (certificates.isEmpty()) {
+            throw new IllegalStateException(
+                    "Must provide at least one certificate to pin network connection to!");
+        }
+
+        for (int i = 0; i < certificates.size(); i++) {
+            String name = provider.getProviderName() + "_" + i;
+            keyStore.setCertificateEntry(name, certificates.get(i));
+        }
+
+        return keyStore;
+    }
+
+    private void init(CertificateProvider provider) {
+        try {
+            keyStore = getKeyStore(provider);
+
+            String kmfAlgorithm = KeyManagerFactory.getDefaultAlgorithm();
+            KeyManagerFactory keyManagerFactory = KeyManagerFactory.getInstance(kmfAlgorithm);
+            keyManagerFactory.init(keyStore, password);
+
+            TrustManagerFactory trustManagerFactory = TrustManagerFactory.getInstance(kmfAlgorithm);
+            trustManagerFactory.init(keyStore);
+
+            SSLContext context = SSLContext.getInstance("TLS");
+            context.init(
+                    keyManagerFactory.getKeyManagers(),
+                    trustManagerFactory.getTrustManagers(),
+                    null
+            );
+
+            sslContext = context;
+            pendingException = null;
+        } catch (GeneralSecurityException | IOException e) {
+            pendingException = e;
```

Comment:
Why defer checking of exceptions? Why not just return a `null` instance of the `SSLManager` for the given provider?

Reply:
Considering this is an integral part in starting a network connection, I though it wise to throw the exact exception (fail hard fast) so that we know exactly where something with SSL went wrong.

I am deferring it so that exceptions are thrown upon use of any `SSLManager` API. This allows us to catch those errors alongside other errors. In this case, that would be the `ButtonApiImpl` class. You'll notice that we have all exception catching already present there.